### ISSUE TITLE
Multi-token greedy decode diverges from llama.cpp across ALL architectures; CompileTraced falls back at the prefill->decode transition

### ISSUE THREAD

**Issue by dndungu** [zerfoo/zerfoo#990]
## Summary

Multi-token greedy decode diverges from llama.cpp after the **first** token, on byte-identical GGUF files, and this reproduces on **every architecture tested — not one builder**. Output stays on-topic English but becomes repetitive.

Observed on `qwen3` (new builder, PR #989) AND on `gemma3`, an accepted, long-shipped architecture, using the same harness. That rules out a per-builder defect and points at the shared generation path.

## Evidence

First-token greedy parity vs llama.cpp on the byte-identical GGUF is largely fine (6/8 on Qwen3-0.6B-Q8_0; the 2 misses are contested continuations where zerfoo is arguably correct). Divergence begins at **token 2**.

Representative output:

```
prompt:  "The capital of France is"
zerfoo:  " Parisian city is the capital of France, and the capital of France is..."
```

Right first token, then it degrades into repetition rather than continuing coherently.

## The lead worth pulling

Every generation logs, at the prefill -> decode transition:

```
CompileTraced plan validation failed, falling back to Compile:
instruction 0 (Mul): input tensors cannot be nil
```

So the traced-execution plan is being **rejected and silently fallen back** on every single decode. Two things are wrong there regardless of the decode quality question: a plan that never validates is dead code paying a validation cost per generation, and a nil input tensor at instruction 0 suggests decode-phase state (KV cache view? position/mask tensor?) is not wired into the traced plan the way prefill's is.

That signature is a plausible root cause for the divergence and should be the first thread pulled. It may also be related to #981 (KV-cache multi-head key offset), which is a different symptom of decode-phase state being wrong.

## Why this matters more than it looks

The verified-model matrix (T136.3) is the Phase 1 exit deliverable, and it asserts generation correctness per model. If multi-token decode is systematically wrong across all architectures, then either the matrix rows will surface it (good, but it blocks the phase) or the parity assertions are too shallow to catch it (worse — same class as the vacuous-parity hazard fixed in #987).

Recommend triaging this **before** T136.3 publishes, so the matrix is not built on a decode path with a known systematic defect.

## Reproduction

`tests/parity/qwen3_test.go` (`QWEN3_GGUF_PATH` set to a Qwen3-0.6B Q8_0 GGUF) and the equivalent gemma3 generation test. Compare against llama.cpp on the same file. Watch stderr for the CompileTraced line.

## Provenance

Found during E151 (Qwen 3 architecture support, PR #989) on 2026-08-20. **Not caused by that work** — confirmed by reproducing on `gemma3` on the same harness. Reported rather than absorbed into the Qwen 3 PR, whose verified-models.md row states the decode limitation explicitly rather than claiming bit-exactness.

**Comment by dndungu** [comment]
## Triaged and fixed — PR #993

Short version: **the `CompileTraced` line is a red herring.** It is a real, unconditional defect worth its own issue, but it is not causing this. The decode divergence is a KV-cache position-offset bug in `GroupedQueryAttention`, and it is a genuine correctness bug — not numerical drift.

All work below was CPU-only on `gemma3-1b-it-Q4_K_M` and `Qwen3-0.6B-Q8_0` from `/var/lib/zerfoo/models`. No GPU needed.

### Root cause

`GroupedQueryAttention` took its RoPE position offset from `CacheProvider.SeqLen()`:

```go
gqa.rope.SetPositionOffset(cache.SeqLen())   // layers/attention/grouped_query_attention.go:774
posOffset = cache.SeqLen()                   // :610, fused QK-norm+RoPE decode path
```

`SeqLen()` reports **layer 0's** cursor in every provider except `GPUKVCache` (`generate/kvcache.go:149`, `kvcache_base.go:151`, `tensor_cache.go:517`, `paged_kv.go:53`, `compressed_kv_cache.go:58`). Layer 0 appends the current chunk to the cache inside its own `Forward` (`grouped_query_attention.go:821`), so layers 1..N-1 in the same pass read a cursor already advanced by `chunkLen`. Logged on a 6-token prompt:

```
[990] layer=0  seqLen=6 cache.SeqLen()=0     <- correct
[990] layer=1  seqLen=6 cache.SeqLen()=6     <- already advanced by layer 0
...
[990] layer=25 seqLen=6 cache.SeqLen()=6
```

Layers 1+ therefore rotate Q/K to positions `[6,12)` instead of `[0,6)`.

**Why first-token parity looked fine.** The shift is *uniform inside one pass*, and RoPE is relative — `q·k` depends only on `pos_q − pos_k`. So a pure prefill and a pure token-at-a-time decode are each internally self-consistent and give correct attention. The damage lands only at the prefill→decode boundary: after an N-token prefill, layers 1+ hold keys at `[N, 2N)` while the first decode token is rotated to `N+1`, and every later query is scored against scrambled relative offsets. Fluent, on-topic, prompt-blind — exactly the reported symptom.

Measured: split the same 6-token prompt as `prefill(k)` + `(6−k)` decode steps, compare last-position logits against a one-shot 6-token prefill.

| split k | maxAbsDiff vs one-shot |
|---|---|
| 5 | **7.448** |
| 4 | **6.532** |
| 3 | **4.636** |
| 2 | **0.844** |
| 1 | 0.000057 |

Error proportional to the prefill chunk, and effectively zero at k=1 where the "prefill" degenerates into pure decode.

### Correctness bug, not drift

Holding this to the bar the issue set. A control that discards the KV cache and re-prefills the whole sequence from scratch at every step is arithmetic the position bug cannot reach. Before the fix, on the identical greedy config, cached decode and that control disagreed from token 2:

```
A incremental-KV : "Paris is a city that is known for"
B full-reprefill : "Paris.\n\nI am a large language"
```

Same weights, same engine, same sampling (`Temperature=0` → the greedy argmax fast path at `generate/generator.go:563`), same process. Two evaluation orders of the same function disagreeing that far is a wiring defect, not float noise. After the fix they are token-identical (`MATCH=true`), and full generation reads:

```
before: "Paris is a city that is known for its beautiful architecture and the French culture."
after:  "Paris.\n\nI am a large language model. I am trained on a massive"
```

### The `CompileTraced` failure — root-caused, and wasteful rather than causal

`embeddingLookupNode.Forward` builds its output in plain Go on the CPU path with **no engine calls** (`inference/arch_llama.go:322-341`). `CompileTraced` builds its slot graph purely from `EngineProxy`-recorded ops (`ztensor@v1.19.2/graph/compile.go:758-779`), so the embedding output has no producing instruction — it only gets a slot when the next traced op consumes it. That slot is neither an input nor a frozen slot, so `RunInstructions` finds it nil and bails at `compile.go:545`: `instruction 0 (Mul)`, the first engine op downstream of the embedding. Structural, not data-dependent — **the traced plan can never validate for any GGUF architecture**, which is why it fires on literally every generation.

Not causal, on three independent lines:

1. The A/B harness above uses `graph.Forward` only and never builds a plan. It reproduced the divergence.
2. Instrumenting the first decode step shows the wrong token is already chosen by the uncompiled `graph.Forward`, **before** `compileGraph` runs — `graph.Forward(reset=false) argmax=563`, and 563 is what was emitted. Compilation's extra forward passes do not clobber the logits about to be sampled.
3. PR #993 changes nothing in the compile or plan path, the `CompileTraced` line still logs on every generation, and decode is now byte-identical to cache-free evaluation. Tokens 3+ in the "after" output are produced *by* the fallback plan, so it cannot be producing wrong arithmetic.

The fallback is also structurally sound: `Compile` reuses `g.memo` from the last real `Forward` instead of re-running it (`compile.go:580-602`), and its instructions call `n.Forward(ctx, ...)` with the runtime context, so the real KV cache is read from `genCtx` at `Run` time.

Still worth its own issue: it costs a full extra forward pass plus a failed validation run on the first decode of every process, permanently disables the traced/megakernel path for every GGUF model, and its log line has now cost at least one investigation its first day. Fix is either to teach the tracer about nodes producing tensors outside the engine, or to skip `CompileTraced` when the graph contains one and drop the log to debug level.

### #981 — related family, different root cause

Not shared. #981 is a buffer-striding defect in `TensorCache`: the buffer is allocated `[batch, maxSeqLen, dim]` (`tensor_cache.go:143`), `Update` writes at the flat offset `lb.seqLen * dim * batch` (`:239`) which only suits a token-major `[maxSeqLen, batch, dim]` layout, and `Get` returns `kBuf[:batch*seqLen*dim]` shaped `[batch, seqLen, dim]` (`:459`) which assumes head-major with a `seqLen*dim` stride. Three layouts, one buffer. Invisible at `batch == 1`, appears as soon as GQA flattens `numKVHeads > 1` into the batch axis — which is precisely the shift the issue reports, so the hypothesis in #981 is correct. `KVCache.Update` already does it right, striding per batch element into `bi*maxSeqLen*dim + cursor*dim` (`kvcache.go:130-135`).

Two corrections for #981: it is **not GPU-specific** — `TestTensorCache_FP16_GPU_MultiHead_PrefillAndDecode` fails on a plain CPU `go test ./generate/`, so no GB10 time is required to fix or verify it. And PR #993 does not fix it; `LayerSeqLen` on `TensorCache` does not touch the striding.

### Fix and blast radius

PR #993 adds `generate.LayerSeqLenProvider` (`LayerSeqLen(layer int) int`) implemented across every cache provider, and routes both GQA offset sites through it with a `Get`-based fallback for out-of-tree providers. `GPUKVCache` was already correct — it advances `seqLen` only after the last layer's append (`gpu_kv_cache.go:165`), the pass-stable design the others should have had.

Regression test `TestGQA_PrefillDecodeRoPEPositionParity` pins `prefill(7) + 3 decode == one-shot prefill(10)` through a 3-layer stack sharing one cache. Red-proofed: maxdiff **0.297** against a 1e-3 tolerance with the old offset, exactly 0 with the new one. A single-layer stack cannot observe this bug at all, which is why the existing `TestGQA_CachedForward` never caught it — it also explicitly declines to compare cached against uncached *values* ("we verify the shapes are correct"), the L-0009 vacuous-assertion pattern. The new test asserts its own sensitivity before it asserts parity, because an earlier draft using the default random init saturated the softmax onto one key, made the stack position-blind, and passed with the bug fully present.

**For T136.3:** every generation zerfoo has produced past the first token was affected. Any golden output, cached parity baseline, or "verified" generation captured before this change is invalid and needs regenerating. Recommend the matrix publish after #993 lands.

Recorded as lore L-0019.

**Comment by dndungu** [comment]
Split the `CompileTraced` half out to #994 so it does not stay bundled with this issue's resolution. This issue is now purely the RoPE position-offset bug, fixed by #993.
